### PR TITLE
Reapply: Mark drop calls in landing pads cold instead of noinline

### LINKED ISSUE #46515: Missing optimization around __rust_alloc and unknown functions if panic=unwind
The following code gets optimized as expected:
```rust
pub fn test() -> bool {
    let x = &*Box::new(0);
    let y = &*Box::new(0);
    
    if x as *const _ == y as *const _ {
        return true;
    }
    return false;
}
```
[becomes](https://godbolt.org/g/fkwdR5)
```
; example::test
; Function Attrs: nounwind uwtable
define zeroext i1 @example::test() unnamed_addr #0 personality i32 (i32, i32, i64, %"unwind::libunwind::_Unwind_Exception"*, %"unwind::libunwind::_Unwind_Context"*)* @rust_eh_personality !dbg !18 {
  ret i1 false, !dbg !22
}
```

However, if I let the code call an unknown function instead of return, the optimization disappears:
```rust
pub fn test(f: fn()) {
    let x = &*Box::new(0);
    let y = &*Box::new(0);
    
    if x as *const _ == y as *const _ {
        f();
    }
}

```
[becomes](https://godbolt.org/g/y5acAz)
```
; example::test
; Function Attrs: uwtable
define void @example::test(void ()* nocapture nonnull %f) unnamed_addr #1 personality i32 (i32, i32, i64, %"unwind::libunwind::_Unwind_Exception"*, %"unwind::libunwind::_Unwind_Context"*)* @rust_eh_personality !dbg !36 {
  %err.i.i.i10 = alloca %"core::mem::ManuallyDrop<alloc::allocator::AllocErr>", align 8
  %_5.i.i.i.i11 = alloca %"alloc::allocator::AllocErr", align 8
  %err.i.i.i = alloca %"core::mem::ManuallyDrop<alloc::allocator::AllocErr>", align 8
  %_5.i.i.i.i = alloca %"alloc::allocator::AllocErr", align 8
  %_9.sroa.9.i.i = alloca [16 x i8], align 8
  %_9.sroa.9.0.sroa_idx9.i.i = getelementptr inbounds [16 x i8], [16 x i8]* %_9.sroa.9.i.i, i64 0, i64 0, !dbg !38
  call void @llvm.lifetime.start(i64 16, i8* nonnull %_9.sroa.9.0.sroa_idx9.i.i), !dbg !38
  %0 = getelementptr inbounds %"core::mem::ManuallyDrop<alloc::allocator::AllocErr>", %"core::mem::ManuallyDrop<alloc::allocator::AllocErr>"* %err.i.i.i, i64 0, i32 0, i64 0, !dbg !48
  call void @llvm.lifetime.start(i64 24, i8* nonnull %0) #6, !dbg !48, !noalias !52
  %1 = call i8* @__rust_alloc(i64 4, i64 4, i8* nonnull %0) #6, !dbg !55, !noalias !52
  %2 = icmp eq i8* %1, null, !dbg !57
  br i1 %2, label %bb3.i.i.i, label %"_ZN35_$LT$alloc..boxed..Box$LT$T$GT$$GT$3new17he6f20be442f24a38E.exit", !dbg !61

[snip]
```

That seems wrong, why should the optimization stop kicking in? The corresponding C++ code does not have any problem:
```C++
#include <memory>

void test(void (*f)()) {
    auto x = std::make_unique<int>(0);
    auto y = std::make_unique<int>(0);

    if (&*x == &*y) {
        f();
    }
}
```
[becomes](https://godbolt.org/g/NqnDcd)
```
; Function Attrs: uwtable
define void @test(void (*)())(void ()* nocapture) local_unnamed_addr #0 personality i8* bitcast (i32 (...)* @__gxx_personality_v0 to i8*) !dbg !1204 {
  tail call void @llvm.dbg.value(metadata void ()* %0, i64 0, metadata !1208, metadata !1211), !dbg !1212
  tail call void @llvm.dbg.value(metadata %"class.std::unique_ptr"* undef, i64 0, metadata !1209, metadata !1213), !dbg !1214
  tail call void @llvm.dbg.value(metadata %"class.std::unique_ptr"* undef, i64 0, metadata !1209, metadata !1213), !dbg !1214
  tail call void @llvm.dbg.value(metadata %"class.std::unique_ptr"* undef, i64 0, metadata !1210, metadata !1213), !dbg !1215
  tail call void @llvm.dbg.value(metadata %"class.std::unique_ptr"* undef, i64 0, metadata !1210, metadata !1213), !dbg !1215
  tail call void @llvm.dbg.value(metadata %"class.std::unique_ptr"* undef, i64 0, metadata !1209, metadata !1213), !dbg !1214
  ret void, !dbg !1216
}
```
(I briefly thought maybe unwinding is the problem, but C++ should have the same kind of unwinding here.)

### LINKED ISSUE #87055: Unable to optimize out branches in drop function in unwind landing pad
In [this code](https://godbolt.org/z/4Wrsb4z5f):
```rust
pub fn foo(g: fn()) {
    let _var: Option<Box<i32>> = None;
    g();
}
```

Assembly:
```asm
core::ptr::drop_in_place<core::option::Option<alloc::boxed::Box<i32>>>:
        mov     rdi, qword ptr [rdi]
        test    rdi, rdi
        je      .LBB0_1
        mov     esi, 4
        mov     edx, 4
        jmp     qword ptr [rip + __rust_dealloc@GOTPCREL]
.LBB0_1:
        ret

example::foo:
        push    rbx
        sub     rsp, 16
        mov     qword ptr [rsp + 8], 0
        call    rdi
        add     rsp, 16
        pop     rbx
        ret
        mov     rbx, rax
        lea     rdi, [rsp + 8]
        call    core::ptr::drop_in_place<core::option::Option<alloc::boxed::Box<i32>>>
        mov     rdi, rbx
        call    _Unwind_Resume@PLT
        ud2

DW.ref.rust_eh_personality:
        .quad   rust_eh_personality
```

`_var` is known to be `None` so the deallocation is optimized away in the normal flow. But in landing pad for unwinding, it still generates a branch to check if `_var` is null. But it's a local variable which is not able to be modified in anywhere else and should be known to be null.

I'm expecting the `drop` of `_var` should be completely elided so that `g()` can be a tail call.

### PR DESCRIPTION
**Pull request by erikdesjardins**
Reapplies #92419, which was reverted in #94402 due to #94390.

Fixes #46515, fixes #87055.

r? @ghost

### CODE REVIEW

File: compiler/rustc_codegen_llvm/src/builder.rs [hunk 1]
```diff
@@ -1188,9 +1189,19 @@ impl<'a, 'll, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
...
-        attributes::apply_to_callsite(llret, llvm::AttributePlace::Function, &[noinline]);
+    fn apply_attrs_to_cleanup_callsite(&mut self, llret: &'ll Value) {
+        let mut attrs = SmallVec::<[_; 2]>::new();
+
+        // Cleanup is always the cold path.
+        attrs.push(llvm::AttributeKind::Cold.create_attr(self.llcx));
+
+        // In LLVM versions with deferred inlining (currently, system LLVM < 14),
+        // inlining drop glue can lead to exponential size blowup, see #41696 and #92110.
+        if !llvm_util::is_rust_llvm() && llvm_util::get_version() < (14, 0, 0) {
```
Comment (nikic) [inline compiler/rustc_codegen_llvm/src/builder.rs:1200]: The is_rust_llvm check shouldn't be necessary.

Comment (erikdesjardins) [reply compiler/rustc_codegen_llvm/src/builder.rs]: Right.

Also, I regret to inform you--it seems like there are more inlining issues.
Compiling https://github.com/MarkDDR/long_compile_rustc_1.59.0 with this patch takes nearly as long as it does with 1.59.0.

**Comment by Dylan-DPC** [comment]
@erikdesjardins any updates on this?

**Comment by erikdesjardins** [comment]
While I concur with nikic's analysis here (https://github.com/rust-lang/rust/issues/94390#issuecomment-1070832077), I still doubt this will be accepted without further LLVM changes.

I guess draft PRs aren't excluded from triage?
@rustbot label S-blocked

**Comment by ldm0** [comment]
@erikdesjardins Hi.

Noticable size bloat(about 2%) and codegen degradation happened on our private project after upgrading rustc. After bisection, I found that the problem was introduced in #94402, which could be resolved by this PR.

I tested locally and found that this PR will fix not only https://github.com/rust-lang/rust/issues/46515, https://github.com/rust-lang/rust/issues/87055, but also #97217 and #98679. As the compile time issue was trimmed(https://github.com/rust-lang/rust/issues/94390#issuecomment-1070832077), could you push this PR forward?

**Comment by erikdesjardins** [comment]
I would also like to see this relanded, but I still agree with my previous comment, so I'm not going to spend the time to update this PR. You have my blessing to rebase these commits and open your own PR if you want.

**Comment by InnovativeInventor** [comment]
> I would also like to see this relanded, but I still agree with my previous comment, so I'm not going to spend the time to update this PR. You have my blessing to rebase these commits and open your own PR if you want.

@erikdesjardins @ldm0 Since there didn't seem to be any activity on this, I took the liberty of rebasing this off master and opening a rebased PR. Let me know how I can help to move this along.

**Comment by Dylan-DPC** [comment]
Closing this in favour of #102099

**Pull request closed without merging**